SAP/spartacus
# feat: add i18n config to lazy-load translations via a JS dynamic import or any Promise

This PR adds the new i18n config `i18n.backend.loader` for lazy-loading translations, allowing for bundling local translations assets via JS dynamic import. For loading **local assets**, this new approach is more performant than to the old config `i18n.backend.loadPath` which caused loading local translations via HTTP - causing HTTP roundtrips in SSR.

The new config property `i18n.backend.loader` is a **function** accepting arguments `langauge` and `namespace` and returning a `Promise` of the translation resources. Instead of making http request, we can now simply use a **dynamic import** to load the translation resources, which will be performant both in CSR and SSR. See the following example:
```ts
{ loader: (language: string, chunkName: string) => import(`../../assets/i18n-assets/${language}/${chunkname}.json`) }
```

Note: We keep the old config for the usecase of loading translations from an **external server** via HTTP.


**CONSEQUENCES OF USING NEW APPROACH**:
- Different filenames of the lazy loaded translations.
    The dynamically-imported JSON translations resources are bundled by webpack and the output files have different filenames and extensions (`.js`) than the original JSON files. 
    In production build the filenames are uglified by default and look for example like below:
    <img width="316" alt="image" src="https://user-images.githubusercontent.com/4001059/216402180-10df7d5e-20e6-4e9e-8736-91ecc5328da4.png">

    In development build the file names look familiar though, for example:
    <img width="304" alt="image" src="https://user-images.githubusercontent.com/4001059/216402373-f981eca1-48d5-40a4-b265-ecc11633426f.png">


closes https://jira.tools.sap/browse/CXSPA-2060

---


## QA steps
Steps to enable lazy-loading in `storefrontapp` (for testing purposes):
- build assets: `npm run build:assets`
- copy i18n-assets to `storefrontapp`: `cp -r dist/assets/i18n-assets projects/storefrontapp/src/assets`
- modify `i18n` config in `app.module.ts`:
```ts
    provideConfig(<I18nConfig>{
      i18n: {
        // resources: translations,
        chunks: translationChunksConfig,
        fallbackLang: 'en',
        backend: {
          loader: (lng, ns) =>
            import(`../assets/i18n-assets/${lng}/${ns}.json`),
        },
      },
    }),
```
- test client-side rendering
  - run the app `npm run start`
  - in the browser open Dev Tools [Network] tab and open the storefront http://localhost:4200
  - verify that:
    1. in the site all texts are translated (no translations missing)
    2. in Network tab see that the translations are now bundled as JS files by webpack:
      <img width="304" alt="image" src="https://user-images.githubusercontent.com/4001059/216402373-f981eca1-48d5-40a4-b265-ecc11633426f.png">
- smoke test SSR:
  - run the ssr dev server: `npm run dev:ssr`
  - [disable javascript](https://developer.chrome.com/docs/devtools/javascript/disable/) in the browser
  - open the server-side-rendered storefront http://localhost:4200 in the browser
  - verify that: in the site all texts are translated (no translations missing)
